### DISCUSSION TITLE
Can I use DapperRepository and EntityRepository in the same transaction?

### DISCUSSION THREAD

**Discussion by CAH-FlyChen** [abpframework/abp#10624] [Q&A]
Can I use DapperRepository and EntityRepository in the same transaction?
eg: 
1. select many items use entity repository get keys
2. use EntityRepository update some business flags
3. use dapper repository to update item in keys batch.
4. commit
because updatemany methord is too slow.
what should i do?

**Comment by enisn** [comment]
**By default, ABP doesn't support that officially.**

Also, I'm not sure how it can be possible. If it's possible, you should implement it by yourself.

**Comment by zhutoutou** [comment]
Although it doesn't support officially, but Dapper and EF can be in same transaction, repo both gets context from UnitOfWorkDbContextProvider and the provider will provide same context in a UOW.  so, it's possible to use dapper and EF in a transaction. 
However, there still remain some problem when we use.

- EF has three periods of data commit 
1. change in memory (Entity change with change tracker or CRUD without autosave)
2. execute SQL (save changes)
3. transaction commit

- Dapper has only two period of data commit 
1. execute SQL
2. transaction commit

Sample:

step 1. EF modifies an entity
step 2. Dapper update field with correlation table
step 3. finally save changes with explicit code or interceptor. 
we may think taht the changes' order in database should be same with code execute flow, but we should know that dapper will execute first in database and EF will apply in database when saving changes.
